src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -293,6 +236,57 @@ Journal::replay_ret CircularBoundedJournal::scan_valid_record_delta(
   });
 }
 
+RecordScanner::read_ret CircularBoundedJournal::read(paddr_t start, size_t len) 
+{
+  LOG_PREFIX(CircularBoundedJournal::read);
+  rbm_abs_addr addr = convert_paddr_to_abs_addr(start);
+  DEBUG("reading data from addr {} read length {}", addr, len);
+  auto bptr = bufferptr(ceph::buffer::create_page_aligned(len));
+  return cjs.read(addr, bptr 
+  ).safe_then([bptr=std::move(bptr)]() {
+    return read_ret(
+      RecordScanner::read_ertr::ready_future_marker{},
+      std::move(bptr)
+    );
+  });
+}
+
+bool CircularBoundedJournal::check_invalid_record_header(scan_valid_records_cursor &cursor,
+  record_group_header_t &r_header) 
+{
+  LOG_PREFIX(CircularBoundedJournal::check_invalid_record_header);
+  auto print_invalid = [FNAME](auto &r_header) {
+    DEBUG("invalid header: {}", r_header);
+    return true;
+  };
+  auto addr = convert_paddr_to_abs_addr(cursor.seq.offset);
+  if (addr + r_header.mdlength + r_header.dlength > get_journal_end() ||
+      r_header.segment_nonce != cjs.get_cbj_header().magic) {
+    return print_invalid(r_header);
+  }
+  return false;
+}
+
+bool CircularBoundedJournal::is_passed_end(scan_valid_records_cursor &cursor,
+  record_group_header_t &r_header) 
+{
+  LOG_PREFIX(CircularBoundedJournal::is_passed_end);
+  auto print_invalid = [FNAME](auto &r_header) {
+    DEBUG("invalid header: {}", r_header);
+    return true;
+  };
+  if (cursor.seq.offset == convert_abs_addr_to_paddr(
+      get_records_start(), get_device_id())) {
```

Comment:
I think we were discusssing how to infer the current segment_seq from `committed_to` for the first record-group, and it is for sure that its `committed_to` must point to a point before rolling, if it has a valid one.

But when it comes to the second/third records, I just realized their current segment sequence still cannot be inferred. It doesn't look correct in the `else if` case that assumes that `committed_to` must be updated to the current sequence.

Note that RecordSubmitter is allowed to submit and write multiple record-groups in parallel, before the previous **ones** are committed to the disk (and so advance the `committed_to`).

Reply:
Can you take a look at the last commit I added? I added not only the following lines but also an unittest.

```
   } else if (r_header.committed_to.segment_seq != cursor.seq.segment_seq) {
+    if (r_header.committed_to == cursor.last_committed) {
+      return false;
+    }
     return print_invalid(r_header);
   } 
```

Reply:
see https://github.com/ceph/ceph/pull/52128#discussion_r1294153920

Reply:
Addressed.